File: PythonCodes/Chapter 10/Fig10_03.py
#%%
"""
Created on Feb 13 2019
Stopping time simulation for Brownian motion
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt

def GeneratePathsBM(NoOfPaths,NoOfSteps,T):    
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    W = np.zeros([NoOfPaths,NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
        
    dt = T / float(NoOfSteps)
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])
        W[:,i+1] = W[:,i] + np.sqrt(dt)*Z[:,i]
        time[i+1] = time[i] +dt
        
    paths = {"time":time,"W":W}
    return paths

def mainCalculation():
    NoOfPaths = 3
    NoOfSteps = 500
    T = 10.0

    # Define the barrier

    B =2.0 
    np.random.seed(2)
    Paths = GeneratePathsBM(NoOfPaths,NoOfSteps,T)
    timeGrid = Paths["time"]
    W = Paths["W"]
    
    plt.figure(1)
    plt.plot(timeGrid, np.transpose(W))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("W(t)")
   
    hittingTime=[]
    for i in range(0,NoOfPaths):
        idx = np.where(W[i,:]>B)
        if idx[0]!=np.array([]):
            plt.plot(timeGrid[idx[0][0]],W[i,idx[0][0]],'or')
            hittingTime.append(timeGrid[idx[0][0]])
            
    plt.figure(2)
    plt.hist(hittingTime,50)
    plt.grid()
mainCalculation()

File: MatlabCodes/Chapter 10/Fig10_03.m
function StoppingTime
close all;clc;
randn('seed',20);
NoOfPaths = 3;
NoOfSteps = 250;

% Random samples

Z = random('normal',0,1,[NoOfPaths,NoOfSteps]);
W = zeros(NoOfPaths,NoOfSteps);

% Maturity time

T = 25;

% Barrier value

B = 2.0;

dt = T/NoOfSteps;
for i = 1: NoOfSteps
    if NoOfPaths>1
        Z(:,i)   = (Z(:,i) - mean(Z(:,i))) / std(Z(:,i));
    end    
    W(:,i+1) = W(:,i) + sqrt(dt) * Z(:,i);
end
time = 0:dt:T;

% Plot the Brownian Motion

figure(1)
plot(time,W,'linewidth',1.5)
grid on; hold on

% Stopping time, i.e. find the time at which the path will hit the barrier

T      =NaN(NoOfPaths,1);
Tind   =NaN(NoOfPaths,1);
PathId =NaN(NoOfPaths,1);
for i =1:NoOfPaths
    ind=find(W(i,:)>=B);
    if ~isnan(ind)
        PathId(i)=i;
        T(i)=time(ind(1));
        Tind(i)=ind(1);
    end
end

a=axis;
legend('path 1','path 2','path 3')
for i=1:length(PathId)
    x=T(i);
    plot(x,2,'.k','markersize',20)
end
plot([0,a(2)],[2,2],'--k')
xlabel('time')
ylabel('W(t)')

% Plot the histogram of the hitting times

figure(2)
hi = histogram(T,25);
hi.FaceColor = [1,1,1];
grid on
xlabel('hitting time')
ylabel('count')
